substrate/frame/migrations/src/lib.rs

```diff
@@ -0,0 +1,750 @@
+	/// - `None`: The migration process is completely finished.
+	/// - `ControlFlow::Break`: Continue in the *next* block with the given cursor.
+	/// - `ControlFlow::Continue`: Continue in the *current* block with the given cursor.
+	fn exec_migration(
+		mut cursor: ActiveCursorOf<T>,
+		is_first: bool,
+		meter: &mut WeightMeter,
+	) -> Option<ControlFlow<ActiveCursorOf<T>, ActiveCursorOf<T>>> {
+		// The differences between the single branches' weights is not that big. And since we do
+		// only one step per block, we can just use the maximum instead of more precise accounting.
+		if meter.try_consume(Self::exec_migration_max_weight()).is_err() {
+			defensive_assert!(!is_first, "There should be enough weight to do this at least once");
+			return Some(ControlFlow::Break(cursor))
+		}
+
+		let Some(id) = T::Migrations::nth_id(cursor.index) else {
+			// No more migrations in the tuple - we are done.
+			defensive_assert!(cursor.index == T::Migrations::len(), "Inconsistent MBMs tuple");
+			Self::deposit_event(Event::UpgradeCompleted);
+			Cursor::<T>::kill();
+			T::MigrationStatusHandler::completed();
+			return None;
+		};
+
+		let Ok(bounded_id): Result<IdentifierOf<T>, _> = id.try_into() else {
+			defensive!("integrity_test ensures that all identifiers' MEL bounds fit into CursorMaxLen; qed.");
+			Self::upgrade_failed(Some(cursor.index));
+			return None
+		};
+
+		if Historic::<T>::contains_key(&bounded_id) {
+			Self::deposit_event(Event::MigrationSkipped { index: cursor.index });
+			cursor.goto_next_migration(System::<T>::block_number());
+			return Some(ControlFlow::Continue(cursor))
+		}
+
+		let max_steps = T::Migrations::nth_max_steps(cursor.index);
+		let next_cursor = T::Migrations::nth_transactional_step(
+			cursor.index,
+			cursor.inner_cursor.clone().map(|c| c.into_inner()),
+			meter,
+		);
+		let Some((max_steps, next_cursor)) = max_steps.zip(next_cursor) else {
+			defensive!("integrity_test ensures that the tuple is valid; qed");
+			Self::upgrade_failed(Some(cursor.index));
+			return None
+		};
+
+		let took = System::<T>::block_number().saturating_sub(cursor.started_at);
+		match next_cursor {
+			Ok(Some(next_cursor)) => {
+				let Ok(bound_next_cursor) = next_cursor.try_into() else {
+					defensive!("The integrity check ensures that all cursors' MEL bound fits into CursorMaxLen; qed");
+					Self::upgrade_failed(Some(cursor.index));
+					return None
+				};
+
+				Self::deposit_event(Event::MigrationAdvanced { index: cursor.index, took });
+				cursor.inner_cursor = Some(bound_next_cursor);
+
+				if max_steps.map_or(false, |max| took > max.into()) {
+					Self::deposit_event(Event::MigrationFailed { index: cursor.index, took });
+					Self::upgrade_failed(Some(cursor.index));
+					None
+				} else {
+					// A migration cannot progress more than one step per block, we therefore break.
+					Some(ControlFlow::Break(cursor))
```

Comment:
Why? If there is still space in the block, why not continue?

Reply:
Yes we could continue here. I will probably change it, there is just the assumption that each migration already does maximal progress when called with a specific weight, so it should be a No-OP to call it again with less weight.